ir/instr.cpp

```diff
@@ -1123,6 +1123,102 @@ unique_ptr<Instr> ExtractValue::dup(const string &suffix) const {
 }
 
 
+void InsertValue::addIdx(unsigned idx) {
+  idxs.emplace_back(idx);
+}
+
+vector<Value*> InsertValue::operands() const {
+  return { val, elt };
+}
+
+void InsertValue::rauw(const Value &what, Value &with) {
+  RAUW(val);
+  RAUW(elt);
+}
+
+void InsertValue::print(ostream &os) const {
+  os << getName() << " = insertvalue " << *val << ", " << *elt;
+  for (auto idx : idxs) {
+    os << ", " << idx;
+  }
+}
+
+StateValue InsertValue::toSMT(State &s) const {
+  auto sv = s[*val];
+
+  Type *type = &val->getType();
```

Comment:
this seems like a lot of repeated code here.
Since indexes are constant, I suggest you extract elements & construct a new value with aggregateVals. You only need to go deep on the indexes path; the remaining ones you keep top-level. i.e. extract idx = 0, append to vector, extract idx=1, but since it's an update, go deeper, etc. then aggregate back.
With a scheme like this there's zero need for new bit-manipulating code, which we don't really want, as it limits opportunities to change representation. And code below is not really possible in this file. Internals like bit-level representation aren't supposed to be exposed in the semantics (this file).

Reply:
true. Following your suggestion, made it much simpler. Thanks. I got inspired by logic in InsertElement::update earlier.